spec/predicates/test-results.md

```diff
@@ -0,0 +1,163 @@
+used the right configurations. If verified using in-toto layouts, a custom
+inspection may optionally parse the `testRun.link` field to verify attestation
+matches the test run.
+
+### Asserting Test Results
+
+In addition to the previous use case, the supply chain owner creates a policy
+verifying that test results passed. In the simplest case, the policy applies to
+all tests. Therefore, it asserts the contents of `result` and that
+`testRun.failedTests` is empty. In more nuanced cases, a subset of tests may
+matter. For example, the tested artifact may be an OS image that's to be
+deployed to three types of devices. As such, the test harness tests the new
+image on an instance of each device. When verifying the attestation prior to an
+image being installed on a device, it only matters that the tests passed on the
+corresponding test device and not necessarily the others.
+
+As before, if the attestation is verified using an in-toto layout, a custom
+inspection may examine the `testRun.link` contents to verify the contents of the
+attestation.
+
+## Prerequisites
+
+Understanding of the
+[in-toto attestation specification](https://github.com/in-toto/attestation).
+
+## Model
+
+This predicate type includes two compulsory fields, `result`, that describes the
+result of the test run, and `testRun` object contains the configuration of the
+tests, as well as optionally a link to the test run, and lists of tests that
+passed, failed, and passed with warnings.  The expected `subject` for this are
+the source artifacts themselves or some reference to them such as a Git commit.
+
+## Schema
+
+```json
+{
+    "_type": "https://in-toto.io/Statement/v1",
+    "subject": [{...}],
+    "predicateType": "https://in-toto.io/attestation/test-result/v0.1",
+    "predicate": {
+        "result": "pass|fail",
+        "testRun": {
+            "link": "<URL>",
+            "configuration": ["<ResourceDescriptor>", ...],
+            "passedTests": ["<TEST_NAME>", ...],
+            "warnedTests": ["<TEST_NAME>", ...],
+            "failedTests": ["<TEST_NAME>", ...]
+        }
+    }
+}
+```
+
+### Parsing Rules
+
+This predicate follows the
+[in-toto Attestation Framework's parsing rules](../v1.0/README.md#parsing-rules).
+
+### Fields
+
+`result` _boolean_ , _required_
+
+Indicates the result of the test run. If true, it indicates _all_ tests passed
+in the corresponding run.
+
+`testRun` _object_, _required_
+
+`testRun.link` _URI_, _optional_
+
+Contains a link to the test run. Useful to point to information that cannot be
+captured in the attestation.
+
+`testRun.configuration` _list of ResourceDescriptor_, _required_
+
+Reference to the configuration used for the test run.
+
+`testRun.passedTests` _list_, _optional_
+
+Each entry corresponds to a single test that passed.
+
+`testRun.warnedTests` _list_, _optional_
+
+Each entry corresponds to a single test that expressed a warning.
```

Comment:
That makes me wonder if instead of having a top level true/false field it would make sense to have top level fields like:

```
testRun.numPassedTests
testRun.numWarnedTests
testRun.numFailedTests
```

Then evaluators are suggested to check:

1. numPassedTests > 0
2. numFailedTests == 0

But can check other stuff if they want too?

Reply:
I think this can be calculated using `len(testRun.{passed,warned,failed}Tests)` correct?

Reply:
Yeah that would work (details would differ depending on language).

If you take this approach would you drop `result`?

Reply:
I think the result is useful as for simple cases, we can drop the passed, warned, failed lists altogether and just include the result field and the configuration.

Reply:
Separately, I wonder if we can move all of the testRun fields to the top level? Now that I think about it, I'm not sure we need that separation.